File: src/memory/memoryReplay.ts
import stringify from "fast-json-stable-stringify";
import type { RuntimeGraph } from "../contracts/graphContracts.js";
import { buildRuntimeMemory, stableMemoryHash } from "./runtimeMemory.js";
import { buildMemoryLineage } from "./memoryLineage.js";
import { buildRuntimeMemoryGraph } from "./runtimeMemoryGraph.js";

export function replayMemoryState(
  graph: RuntimeGraph,
  history: unknown[] = [],
): Record<string, unknown> {
  const mem = buildRuntimeMemory(graph, history);
  const lineage = buildMemoryLineage(
    history.map((h, i) => (typeof h === "object" && h ? { ...(h as object), tick: i } : { tick: i })),
  );
  const memory_graph = buildRuntimeMemoryGraph(graph, history);
  return {
    ...mem,
    lineage: lineage.lineage,
    memory_graph,
    replay_hash: stableMemoryHash(graph, history),
    bounded: true,
  };
}

export function validateMemoryReplay(
  original: Record<string, unknown>,
  replayed: Record<string, unknown>,
): boolean {
  return stringify(original.stable_hash) === stringify(replayed.stable_hash);
}

File: lib/src/memory/memory_replay.dart
import '../graph/runtime_graph.dart';
import 'memory_lineage.dart';
import 'runtime_memory.dart';
import 'runtime_memory_graph.dart';

Map<String, dynamic> replayMemoryState(
  RuntimeGraph graph, [
  List<dynamic> history = const [],
]) {
  final mem = buildRuntimeMemory(graph, history);
  final histMaps = history
      .map((h) => h is Map<String, dynamic> ? h : {'tick': history.indexOf(h)})
      .cast<Map<String, dynamic>>()
      .toList();
  final lineage = buildMemoryLineage(histMaps);
  final memoryGraph = buildRuntimeMemoryGraph(graph, history);
  return {
    ...mem,
    'lineage': lineage['lineage'],
    'memory_graph': memoryGraph,
    'replay_hash': stableMemoryHash(graph, history),
    'bounded': true,
  };
}

bool validateMemoryReplay(
  Map<String, dynamic> original,
  Map<String, dynamic> replayed,
) =>
    original['stable_hash'] == replayed['stable_hash'];
